### PR TITLE
restore: randomly distribute work during seperate process tenant restore

### PR DESCRIPTION
**Pull request by msbutler**
Previously, the restore split and scatter processsor assumed that each nodeID returned by scatter mapped identically to a sql instance id. This NodeID is used to route work to a downstream restore data processor. This assumption does not hold when a seperate process tenant runs restore.

This hot fix instead randomly chooses a sql instance id to route the chunk to for further processing. A long term fix should remote the split and scatter processor all together, as this hot fix will not obey execution locality constraints during a seperate process restore.

Informs #127924

Release note: none

### CODE REVIEW

File: pkg/ccl/backupccl/generative_split_and_scatter_processor.go [hunk 1]
```diff
@@ -601,6 +601,23 @@ func runGenerativeSplitAndScatter(
 					}
 				}
 
+				if _, ok := flowCtx.NodeID.OptionalNodeID(); !ok {
+					// If a sperate process tenant is running restore, the nodeID returned
```
Comment (stevendanna) [inline pkg/ccl/backupccl/generative_split_and_scatter_processor.go:605]:
```suggestion
					// If a separate process tenant is running restore, the nodeID returned
```

File: pkg/ccl/backupccl/generative_split_and_scatter_processor.go [hunk 2]
```diff
@@ -601,6 +601,23 @@ func runGenerativeSplitAndScatter(
...
+					// by scatter does not map identically to a sqlInstanceID; thus, the
+					// processor must randomly choose a sqlInstanceID to route the chunk
+					// to.
+					//
+					// TODO(msbutler): it is unfortunate that all logic after scatter
+					// operates on a NodeID object. The logic should use
+					// SQLInstanceID's as these chunks are routed to dist sql processors
```
Comment (stevendanna) [inline pkg/ccl/backupccl/generative_split_and_scatter_processor.go:612]:
```suggestion
					// SQLInstanceIDs as these chunks are routed to DistSQL processors
```

**Review by stevendanna — APPROVED** [review]
Thanks for digging into this.

Do we know what the previous behavior of this code was? Did everything end up on the gateway node? Might be nice to include in the commit message.

**Comment by msbutler** [comment]
> Do we know what the previous behavior of this code was? Did everything end up on the gateway node? Might be nice to include in the commit message.

@stevendanna during a mismatch, work was routed to the default sql instance id:
https://github.com/msbutler/cockroach/blob/butler-randomize-tenant-scatter/pkg/ccl/backupccl/restore_processor_planning.go#L130 

Added this to the commit message.

**Comment by msbutler** [comment]
TFTR! 

bors r=stevendanna

**Pull request merged**